src/main/java/software/amazon/awssdk/eventstreamrpc/OperationContinuationHandler.java

```diff
@@ -189,128 +199,155 @@ final public CompletableFuture<Void> sendStreamEvent(final StreamingResponseType
                     handleStreamEvent(streamEvent);
                 }
-            } else { //this is the initial request
+            } else {
+                //this is the initial request
                 initialRequestHeaders = new ArrayList<>(list);
                 initialRequest = serviceModel.fromJson(getRequestClass(), bytes);
                 //call into business logic
-                final ResponseType result = handleRequest(initialRequest);
-                if (result != null) {
-                    if (!getResponseClass().isInstance(result)) {
-                        throw new RuntimeException("Handler for operation [" + getOperationName()
-                                + "] did not return expected type. Found: " + result.getClass().getName());
+                CompletableFuture<ResponseType> resultFuture = handleRequestAsync(initialRequest);
+                if (resultFuture == null) {
+                    resultFuture = CompletableFuture.completedFuture(handleRequest(initialRequest));
+                }
+                resultFuture.handle((result, throwable) -> {
+                    if (throwable != null) {
+                        handleAndSendError(throwable);
+                        return null;
                     }
-                    sendMessage(result, !isStreamingOperation()).whenComplete((res, ex) -> {
-                        if (ex != null) {
-                            LOGGER.error(ex.getClass().getName() + " sending response message: " + ex.getMessage());
-                        } else {
-                            LOGGER.trace("Response successfully sent");
+                    if (result != null) {
+                        if (!getResponseClass().isInstance(result)) {
+                            throw new RuntimeException("Handler for operation [" + getOperationName()
+                                    + "] did not return expected type. Found: " + result.getClass().getName());
                         }
-                    });
-                    invokeAfterHandleRequest();
-                } else {
-                    //not streaming, but null response? we have a problem
-                    throw new RuntimeException("Operation handler returned null response!");
-                }
+                        sendMessage(result, !isStreamingOperation()).whenComplete((res, ex) -> {
+                            if (ex != null) {
+                                LOGGER.error(ex.getClass().getName() + " sending response message: " + ex.getMessage());
+                            } else {
+                                LOGGER.trace("Response successfully sent");
+                            }
+                        });
+                        invokeAfterHandleRequest();
+                    } else {
+                        //not streaming, but null response? we have a problem
+                        throw new RuntimeException("Operation handler returned null response!");
```

Comment:
This runtime used to be synchronous from the `onContinuationMessage` callback, but now it happens as part of  CompletableFuture.handle. I'm not sure where these exceptions go. Do they just vanish? Was this something we expected to catch from whatever calls `onContinuationMessage`? Or did that exception just vanish as well?

Reply:
They were being caught and handled in the place that I extracted to the handleAndSendError. They are still handled that way, and in fact these exceptions aren't really possible anyway.